M2/Macaulay2/m2/intersect.m2

```diff
@@ -29,37 +29,34 @@ eliminationInfo Ring := (cacheValue symbol eliminationInfo) (R -> (
 -- General intersect method
 -----------------------------------------------------------------------------
 
-intersect List     :=
-intersect Sequence := -* [same as input type] => *- true >> opts -> L -> (
-    if not #L > 0 then error "intersect: expected at least one object";
+-- intersect is now declared as a binary associative method in shared.m2,
+-- so the methods (intersect, Sequence) and (intersect, List) are pre-installed.
+-- Here, we override (intersect, List) with a specialized method for handling a uniform list.
+-- As a backup, at the end the List is converted to a Sequence.
+intersect List := true >> opts -> L -> (
```

Comment:
As the comment says, this is to allow for overriding the method `(intersect, List)`, specifically because there are sometimes faster algorithms for intersecting N objects at the same time rather than one by one. This is the case for modules, for instance.

Reply:
I don't understand your answer, because N objects can be provided in a sequence and in a list equally well.

Reply:
No, there's a subtle difference between a list and a sequence in this context: #1878.

More to the point, when the input to an associative binary method is a Sequence it implies "compute by sequentially applying the binary method from left to right," but when a List is given, I wanted to keep the possibility of allowing specialized algorithms that consider the whole list in parallel rather than sequentially.

Note that I specifically don't want to always use the specialized algorithms, because in some cases it may be slower, so it's good to allow both so the user can decide which to use based on the type of input.

Reply:
Let me know if this is sufficient explanation. If you have any suggestions for how to allow a binary method on both lists and sequences and allow both sequential computation and all at once computation, let me know! This was the best solution I could think of.

Reply:
> No, there's a subtle difference between a list and a sequence in this context: #1878.
> 
> More to the point, when the input to an associative binary method is a Sequence it implies "compute by sequentially applying the binary method from left to right," but when a List is given, I wanted to keep the possibility of allowing specialized algorithms that consider the whole list in parallel rather than sequentially.

It seems strange to signal the preferred strategy by the having user pass a list instead of a sequence.  Since sequences are problematic(#1878), why not just insist that lists be used instead of sequences, and offer a "Strategy" option?

Reply:
I believe what you're saying is incompatible with `(tensor, List)` being a binary method. If something like #1979 was implemented, then I could make Lists of modules an exception and do it your way, but I don't know how to implement it another way so that tensoring a list of Modules happens in parallel but a list of rings happens associatively.

Reply:
We could let all such associative methods be iterative by default, and let users specify the simultaneous strategy with the `Strategy` option.  (There doesn't seem to be a burning need for simultaneous tensor products of modules by default.)

Reply:
Ah sorry I meant intersect, not tensor (hence the section of code we're discussing). Using iterative intersection for modules is significantly slower, so I don't see what would be the improvement, and also I don't understand your suggestion with strategy. Where would you check for the strategy? Are you suggesting having all binary associative methods take options?

I still think having a meaningful and documented distinction between `(intersect, List)` and `(intersect, Sequence)` makes the most sense. You can apply the same logic to other binary functions if you have a better algorithm for simultaneous computation.

Reply:
I don't think there can be a meaningful distinction between `(intersect,List)` and `(intersect,Sequence)`.  Lists and sequences are meant to be basically interchangeable as repositories of objects.

Not all associative methods would need to take options -- just those where you want to offer a way of applying the operation simultaneously to all the arguments.

Reply:
I think I offered a meaningful distinction between the two earlier:
> when the input to an associative binary method is a Sequence it implies "compute by sequentially applying the binary method from left to right," but when a List is given, I wanted to keep the possibility of allowing specialized algorithms that consider the whole list in parallel rather than sequentially.

Otherwise why have both lists and sequences? A lot of code would be much simpler if one of them simply inherited from the other.

Reply:
The reason to have both sequences and lists is to provide something lightweight for use when calling functions with multiple arguments.  A sequence doesn't contain a pointer to its class (so you can't make a new type of sequence), and it also doesn't contain a precomputed hashcode.  Other than that, lists and sequences are meant to be basically interchangeable as repositories of objects, and trying to impose a semantic distinction, like the one you desire, would lead to confusion, even if documented.

Reply:
I don't have an exact change to your code to propose.

Reply:
A suggestion on where to check the value of the strategy option that allows both sequential and simultaneous computation strategies while keeping intersect a binary method would suffice. I've thought about this carefully and I can't think of any possibility that doesn't come with a significant performance hit.